Notification for starting and closing of assessment.

custode opened the pull request:
Send an opening notification when an assessment is starting and a closing notification when an assessment is closing one day before the end day.

Notification will not be send out for draft assessment, and closing notification will not be send out if the assessment is guided (following the previous implementation).

In the event that the start date or end date is changed, notification will only be send on the latest start/end date. E.g. if assessment start date change from 1st nov to 3rd nov, opening notification will only be send out on 3rd nov.

allenwq commented on app/models/course/assessment.rb:
```diff
@@ -110,6 +112,21 @@ def set_defaults
+  def setup_opening_reminders
+    execute_after_commit do
+      Course::Assessment::OpeningReminderJob.set(wait_until: start_at).
+        perform_later(creator, self, start_at.to_i)
```
Think we should use `to_f` and add and add some random milliseconds when the user changes the time. It's possible that that user changes time and then changes back, normally the user will leave the "seconds" part on the UI to zero.

custode replied:
Fixed.

allenwq commented on app/models/course/assessment.rb:
```diff
@@ -13,6 +13,9 @@ class Course::Assessment < ActiveRecord::Base
   before_validation :propagate_course
   before_validation :assign_folder_attributes
+  before_validation :prevent_reminder_duplication
+  before_save :setup_opening_reminders, if: :start_at_changed?, unless: :draft?
```
Should be `after_save` ? so when there's a saving failure the notification does not get sent.

custode replied:
The start_at_changed cannot be detected in after_save, due to acts_as. Before_save is used in this case to register the changes, and job will only be queue after_commit.

There's one more layer of check in the job itself, which ensure that the input start_at date is consistent with the model start_at date.

allenwq commented on app/services/course/assessment/reminder_service.rb:
```diff
@@ -0,0 +1,20 @@
+
+  def closing_reminder(user, assessment, end_at)
+    return unless end_at == assessment.end_at.to_f
+    return if assessment.draft? || assessment.guided?
```
Any reason why guided closing reminder is turned off ?

custode replied:
In v1, is the closing reminder only for mission due?

allenwq commented:
@custode  In v1 there's a summary email to the lecturer too, it ensures that the lecturer knows the reminder emails are sent to their students. Would be good to implement it as well, but can be another PR.

allenwq commented on app/models/course/assessment.rb:
```diff
@@ -110,6 +113,27 @@ def set_defaults
 
+  def prevent_reminder_duplication
+    # Randomize the milliseconds of the reminders' datetime to prevent duplication
+    self.start_at += Random.rand(0...1.0).round(1) if start_at_changed?
```
Shall we make this even smaller, like `Random.rand(0...0.1).round(4)` ?  `round(1)` only returns 10 different values.

allenwq commented:
@custode The tests are failing, I suspect it's because the change of the adapter.

custode commented:
@allenwq You are right that the test are failing due to the change of adapter. Some of our active job tests relies on the modified background thread adapter's logic, which caused it to stuck after enqueuing. I have reverted back to our original adapter and added some additional logic to support the future job.

allenwq commented:
@custode Thanks for the updating !

Sorry, I realized there's one more issue: the front end form does not reflect the floating part, so that every update to assessment now will trigger a job ( because `start_at_changed?` or `end_at_changed?` are always true ). You might need to somehow let rails know that the dates are not changed in that case.

custode commented:
@allenwq Thanks! Added the logic to reset the changes if the only differences are the floating points.

allenwq commented:
@custode Can you add a feature spec for it ?  To ensure that the email is not sent when start_at/end_at are not changed in the update.

custode commented:
@allenwq Added the feature spec for it.

allenwq commented:
@custode Thanks !

weiqingtoh commented on config/locales/en/notifiers/course/assessment_notifier/closing/user_notifications/email.yml:
```diff
@@ -0,0 +1,10 @@
+        closing:
+          user_notifications:
+            email:
+              subject: '%{course} Reminder for about %{assessment}'
```
`%{course}: Reminder on %{assessment}`
